src/ToricVarieties/CohomologyClasses/special_attributes.jl

```diff
@@ -0,0 +1,111 @@
+########################
+# 1: Special attributes of toric varieties
+########################
+
+@doc Markdown.doc"""
+    cohomology_ring(v::AbstractNormalToricVariety)
+
+Return the cohomology ring of the simplicial and complete toric variety `v`.
+
+# Examples
+```jldoctest
+julia> p2 = projective_space(NormalToricVariety, 2);
+
+julia> ngens(cohomology_ring(p2))
+3
+```
+"""
+function cohomology_ring(v::AbstractNormalToricVariety)
+    if ((issimplicial(v) == false) || (iscomplete(v) == false))
+        throw(ArgumentError("The cohomology ring is (currently) only supported for simplicial and complete toric varieties."))
+    end
+    R, _ = PolynomialRing(coefficient_ring(v), coordinate_names(v), cached = false)
+    weights = [1 for i in 1:ngens(R)]
+    R = grade(R, weights)[1]
+    linear_relations = ideal_of_linear_relations(R, v)
+    stanley_reisner = stanley_reisner_ideal(R,v)
+    return quo(R, linear_relations + stanley_reisner)[1]
+end
+export cohomology_ring
+
+
+@doc Markdown.doc"""
+    volume_form(v::NormalToricVariety)
+
+Construct the volume form of the normal
+toric toric variety `v`.
+
+# Examples
+```jldoctest
+julia> polynomial(volume_form(projective_space(NormalToricVariety, 2)))
+x3^2
+
+julia> polynomial(volume_form(del_pezzo(3)))
+-e2^2
+
+julia> polynomial(volume_form(hirzebruch_surface(5)))
+1//5*x2^2
+```
+"""
+@attr CohomologyClass function volume_form(v::NormalToricVariety)
+    mc = ray_indices(maximal_cones(v))
+    exponents = [fmpz(mc[1,i]) for i in 1:length(mc[1,:])]
+    indets = gens(cohomology_ring(v))
+    poly = fmpz(1) * prod([indets[k]^exponents[k] for k in 1:length(exponents)])
+    if (iszero(poly) || degree(poly)[1] != dim(v))
+        throw(ArgumentError("The volume class does not exist."))
+    end
+    return CohomologyClass(v, poly)
+end
+export volume_form
+
+
+@attr function _intersection_form_via_exponents(v::NormalToricVariety)
+    # extract the cohomology classes corresponding to the torus-invariant prime divisors
+    generators = [CohomologyClass(d) for d in torusinvariant_prime_divisors(v)]
+    
+    # find combinations of those classes that we have to integrate
+    S, _ = PolynomialRing(QQ, ["g$(i)" for i in 1:length(generators)], cached=false)
+    S = grade(S, [1 for i in 1:ngens(S)])[1]
+    hc = homogeneous_component(S, [dim(v)])
+    monoms = [hc[2](x) for x in gens(hc[1])]
+    combinations = reduce(vcat,[[[fmpz(l) for l in k] for k in exponent_vectors(m)] for m in monoms])
+    
+    # perform the integrals
+    intersection_dict = Dict{fmpz_mat, fmpq}()
+    for expos in combinations
+        cc = prod([generators[k]^expos[k] for k in 1:length(generators)])
```

Comment:
Better to use a generator expression than a list comprehension; this reduces memory allocation.
```suggestion
        cc = prod(generators[k]^expos[k] for k in 1:length(generators))
```

Just to showcase the difference:
```
julia> @btime prod([i^(i+1) for i in 1:5])
  46.385 ns (1 allocation: 96 bytes)
10368000000

julia> @btime prod(i^(i+1) for i in 1:5)
  15.489 ns (0 allocations: 0 bytes)
10368000000
```

Reply:
Thank you Max!